src/vset.h

```diff
@@ -0,0 +1,317 @@
+ *
+ * Utilities:
+ *     bool vsetIsEmpty(vset *set);
+ *
+ * Iteration:
+ *     void vsetStart(vset *set, vsetIterator *it);
+ *     bool vsetNext(vsetIterator *it, void **entryptr);
+ *     void vsetStop(vsetIterator *it);
+ *
+ *-----------------------------------------------------------------------------
+ * Iteration Support
+ *-----------------------------------------------------------------------------
+ *
+ * Iterator structure maintains context across all bucket types:
+ *
+ *     typedef struct vsetIterator {
+ *         raxIterator riter;           // for RAX
+ *         hashtableIterator hiter;     // for HT
+ *         uint32_t viter;              // for VECTOR
+ *         void *vsingle;               // for SINGLE
+ *         vsetBucket *parent_bucket;   // owning bucket
+ *         vsetBucket *bucket;          // active bucket
+ *         void *entry;                 // current entry
+ *         long long bucket_ts;         // for RAX
+ *         int iteration_state;         // internal FSM
+ *     } vsetIterator;
+ * */
+
+#define VOLATILESET_BUCKET_INTERVAL_MAX (1LL << 13LL) // 2^13 = 8192 milliseconds
+#define VOLATILESET_BUCKET_INTERVAL_MIN (1LL << 4LL)  // 2^4 = 16 milliseconds
+
+#define VOLATILESET_VECTOR_BUCKET_MAX_SIZE 127
+
+typedef long long (*vsetGetExpiryFunc)(const void *entry);
+typedef int (*vsetExpiryFunc)(void *entry, void *ctx);
+
+// Generic bucket type
+typedef void vsetBucket;
+
+// vset is just a pointer to a bucket
+typedef vsetBucket *vset;
+
+typedef struct vsetIterator {
+    /* for rax bucket */
+    raxIterator riter;
+    union {
+        /* for hashtable bucket */
+        hashtableIterator hiter;
+        /* for vector bucket */
+        uint32_t viter;
+        /* for single bucket */
+        void *vsingle;
+    };
+    /* the parent of the bucket we are currently iterating on */
+    vsetBucket *parent_bucket;
+    /* the bucket we are currently iterating on */
+    vsetBucket *bucket;
+    /* the pointer entry */
+    void *entry;
+    /* In case of rax encoded set, this is the current iterated bucket timestamp */
+    long long bucket_ts;
+    /* the state of the iteration */
+    int iteration_state;
+} vsetIterator;
+
+bool vsetAddEntry(vset *set, vsetGetExpiryFunc getExpiry, void *entry);
+bool vsetRemoveEntry(vset *set, vsetGetExpiryFunc getExpiry, void *entry);
+bool vsetUpdateEntry(vset *set, vsetGetExpiryFunc getExpiry, void *old_entry, void *new_entry, long long old_expiry, long long new_expiry);
+bool vsetIsEmpty(vset *set);
+void vsetStart(vset *set, vsetIterator *it);
+bool vsetNext(vsetIterator *it, void **entryptr);
+void vsetStop(vsetIterator *it);
+void vsetInit(vset *set);
+void vsetClear(vset *set);
+long long vsetEstimatedEarliestExpiry(vset *set, vsetGetExpiryFunc getExpiry);
+size_t vsetPopExpired(vset *set, vsetGetExpiryFunc getExpiry, vsetExpiryFunc expiryFunc, mstime_t now, size_t max_count, void *ctx);
```

Comment:
This API is sending mixed messages.
* "popExpired" tells me that we're going to remove and return expired items.
* "expiryFunc" tells me that we're going to call a function on expired items as they are removed.

Seems odd to do both.  If we're going to return the items, the caller can invoke whatever function it wants.  But if we're going to fire a callback on items as we're expiring them, it seems like the callback should be responsible for freeing the associated entry - and then there'd be nothing to return.

Reply:
You are right. We can consider removing the expire callback. It was evolving here and now it mainly serves as some halt optional thing.

Reply:
I made the change. although I am not happy with it. We wanted to avoid re-access memory regions over and over again. this way enabled us to do much (if not all) of the work related to entry expiration:
1. pop from the hashtable
2. feed to the replication stream
3. free the entry while the entry data in in the cache. 

I will take another look and considerif to change it back

Reply:
@JimB123 thought about it some more and decided to revert back to the current state where we get a callback expiry function. the reason is as I mentioned earlier: I prefer to try and maintain cache locality around entries memory. providing an array to host the "popped" entries forces the user to iterate over them and re-access the entries memory while most of the operations around an entry can be done in this function callback.
I will be happy to discuss later (maybe after RC1?)